### ISSUE TITLE
Misleading doc : "single source of truth for computing for how all distances should be calculated"

### ISSUE THREAD

**Issue by Vincent-SV** [TDAmeritrade/stumpy#825]
Hello and thank you for your work on matrix profiles !

I am interested in the technique used in this paper : https://core.ac.uk/download/pdf/188644494.pdf for reducing the influence of noise when calculating the matrix profile.
The only change to the algorithm is to modify the calculation of distances between 2 subsequence (correcting the distance using the variance of the subsequences)

After digging in the code, I found the function core._calculate_squared_distance with the doc saying "Compute a single squared distance given all scalar inputs. This function serves as the single source of truth for how all distances should be calculated."
This function is the exact calculation I want to modify, so I tested a modification by replacing the output by 1

I then run the code below :

    import stumpy
    import numpy as np

    time = np.linspace(-20, 20, 1000)
    data = np.cos(time)

    mp = stumpy.stump(data, m=50)
    motifs = stumpy.motifs(data, mp[:, 0], max_matches=5)
    print(mp)
    print(motifs)

Expected : After the modification, when I run the code above, I should get a constant matrix profile at 1

Instead : The matrix profile is unchanged. But weirdly, the output of motifs IS changed, even though I thought it wouldn't recalculate any distances.

The issue : the doc saying "This function serves as the single source of truth for how all distances should be calculated." is no longer true  (I suspect it has to do with the parameter p=2.0), some changes should be made about that.
Moreover, there seem to be some inconsistency between the source of truth for distance calculation between functions "stump" and "motifs" when p=2.0 (default)

I think a single source of truth for the calculation of distances is something really nice, because I and others (and maybe you) may want to play with different distances (already done with parameter p ?) and a single source of truth makes it easier and reduces possible bugs

**Comment by Vincent-SV** [comment]
Also I get that I am speaking about internal documentation and internal implementation details, so it is totally ok to dismiss this issue.

**Comment by seanlaw** [comment]
@Vincent-SV Thank you for your question and welcome to the STUMPY community

> the doc saying "This function serves as the single source of truth for how all distances should be calculated." is no longer true

Unfortunately, this statement is outdated and needs to be updated (i.e., it is no longer the single source of truth). As a result of:

1. implementing newer (faster) algorithms for computing the matrix profile
2. accommodating other hardware like GPUs (i.e., for programming paradigm reasons, the same z-norm Euclidean distance is computed quite differently on GPUs vs CPUs)
3. adding support for non-normalized Minkowski distance (`normalize=False`, `p >= 0`)

we were forced to separate the logic for computing various distances. Fundamentally, beyond the original STAMP algorithm (leveraging FFT) for computing matrix profiles, the subsequent speed gains are derived from somewhat esoteric optimizations/tricks that require further specialization. While it would have been nice to maintain a single source of truth for computing the z-norm Euclidean distance, this was not sustainable as new features were added to STUMPY. 

Unfortunately with these new features, we were forced to choose between:

1. allowing easy modifications of how the distance is computed but at the expense of decreased speed
2. prioritizing speed/efficiency for computing the z-norm Euclidean distance while decreasing the ease of modification

Alas, given that 99% of our users either don't care to modify how the distances are computed (beyond specifying the input parameters), we chose to cater to that audience and prioritize speed/efficiency. In fact, originally, we only supported z-norm Euclidean distance only and, after receiving more user feedback, we eventually added non-normalized Minkowski distance, which is not part of the original matrix profile papers and, if I were to guess, likely makes up less than ~5% of our users.

> The matrix profile is unchanged. But weirdly, the output of motifs IS changed, even though I thought it wouldn't recalculate any distances.

So, `stumpy.stump` does not call `core._calculate_squared_distance` anymore (it use to, a long time ago). Again, we changed over to a new approach that leverages Pearson correlations and diagonal squared matrix traversal (rather than sequential row-wise traversal) that significantly reduces the number of FLOPs and decreases the computation by 20-30%. This optimization is only relevant when computing the full `n x n` matrix profile. However, for the occasional distance profile or `stumpy.motifs` computation, `core._calculate_squared_distance` may be invoked. One thing of note is that floating point precision/imprecision may be a problem in some cases where the data has large local fluctuations, though, this is exceptionally rare and typically only arises in contrived scenarios.

> I am interested in the technique used in this paper : https://core.ac.uk/download/pdf/188644494.pdf for reducing the influence of noise when calculating the matrix profile.

In case it matters, there is an [open issue](https://github.com/TDAmeritrade/stumpy/issues/127) for this. However, there are many other things we need to do first before we can begin tackling the topic of "noise" and how it may affect the user-facing API (i.e., it may completely out of scope).  

> Also I get that I am speaking about internal documentation and internal implementation details, so it is totally ok to dismiss this issue.

I appreciate that! I know that it doesn't help you with handling noisy data but I will at least address this issue by removing the misleading docstring. Your feedback is helpful!

**Comment by seanlaw** [comment]
@Vincent-SV Please feel free to comment or re-open this issue (it was closed by the PR) if needed

**Comment by Vincent-SV** [comment]
Thanks a lot for the quick and extensive feedback !

I was indeed expecting that it was just outdated internal documentation, sorry to bother you with that.

Also thanks a lot for the pointer to the relevant issue, I looked for it a bit but I guess I just missed it. It hints to the correct changes to make so I am very happy with that.

**Comment by seanlaw** [comment]
> Also thanks a lot for the pointer to the relevant issue, I looked for it a bit but I guess I just missed it. It hints to the correct changes to make so I am very happy with that.

Please feel free to leave any comments/thoughts there if you think there is anything that we are missing
